### ISSUE TITLE
Possible memory leak on to_rgba8

### ISSUE THREAD

**Issue by fairbairn** [ggez/ggez#958]
Grabbing a snapshot of canvas seems to create a memory leak over time, as if something internal to the function is not released.

The loss seems to be around 50-60k per second.

We stripped back our use case to the following...
update does nothing
draw does the following..

```rust
    fn draw(&mut self, ctx: &mut Context) -> GameResult {
        self.capture(ctx);
        self.render(ctx);
        graphics::present(ctx);
        Ok(())
    }
```

We have the following struct..our state which holds a stage.. the state maintains the canvas that the stage draws into..
We grab a capture of that canvas before allowing it to draw..
If we don't perform the capture function, everything works  If we do, we leak memory.

```rust
pub struct State {
    pub stage: Stage,
}

impl State {
    pub fn new(ctx: &mut Context) -> State {
        let canvas = Canvas::with_window_size(ctx).unwrap();

        State {            
            canvas,                  // off-screen canvas
            stage: Stage::new(ctx), 
        }
    }
    
    fn capture(&mut self, ctx: &mut Context) {
        &self.stage.capture_stage(ctx, &self.canvas);
    }

    /// renders the off-screen canvas to the on-screen canvas
    fn render(&self, ctx: &mut Context) {
        let params = DrawParam::new().dest([0.0, 0.0]);
        graphics::set_canvas(ctx, None);
        graphics::draw(ctx, &self.canvas, params).unwrap();
    }
}

pub struct Stage {
    capture_video: Arc<RwLock<Vec<u8>>>,
}

impl Stage {
    pub fn new(ctx: &mut Context) -> Stage {
        let canvas = graphics::Canvas::with_window_size(ctx).unwrap();
        let capture_video = Arc::new(RwLock::new(canvas.image().to_rgba8(ctx).unwrap()));

        Stage {
            capture_video,
        }
    }

    pub fn capture_stage(&mut self, ctx: &mut Context, canvas: &graphics::Canvas) {
        let screen = canvas.image().to_rgba8(ctx).unwrap();    // <<<<<<<<<<<<<<<<<<<<<<<<< THIS CAUSES A LEAK

        // let mut bytes = self.capture_video.write().unwrap();
        // *bytes = screen;
        // drop(bytes);

    }
}
```

This should just work all day long, not really do anything, and `screen` should drop the used memory.

If we allow the `let screen = canvas.image().to_rgba8(ctx).unwrap();` to run without our update or draw event, then memory leaks continuously.

**Hardware and Software:**
 - ggez version:0.6.rc1
 - OS: Big Sir (mac osx)

This is the current source for the to_rgba8() fn...

```rust
   /// Dumps the `Image`'s data to a `Vec` of `u8` RGBA values.
    pub fn to_rgba8(&self, ctx: &mut Context) -> GameResult<Vec<u8>> {
        use gfx::memory::Typed;
        use gfx::traits::FactoryExt;

        let gfx = &mut ctx.gfx_context;
        let w = self.width;
        let h = self.height;

        // Note: In the GFX example, the download buffer is created ahead of time
        // and updated on screen resize events. This may be preferable, but then
        // the buffer also needs to be updated when we switch to/from a canvas.
        // Unsure of the performance impact of creating this as it is needed.
        // Probably okay for now though, since this probably won't be a super
        // common operation.
        let dl_buffer = gfx
            .factory
            .create_download_buffer::<[u8; 4]>(w as usize * h as usize)?;

        let factory = &mut *gfx.factory;
        let mut local_encoder = GlBackendSpec::encoder(factory);

        local_encoder.copy_texture_to_buffer_raw(
            &self.texture_handle,
            None,
            gfx::texture::RawImageInfo {
                xoffset: 0,
                yoffset: 0,
                zoffset: 0,
                width: w as u16,
                height: h as u16,
                depth: 0,
                format: gfx.color_format(),
                mipmap: 0,
            },
            dl_buffer.raw(),
            0,
        )?;
        local_encoder.flush(&mut *gfx.device);

        let reader = gfx.factory.read_mapping(&dl_buffer)?;

        // intermediary buffer to avoid casting.
        let mut data = Vec::with_capacity(self.width as usize * self.height as usize * 4);
        // Assuming OpenGL backend whose typical readback option (glReadPixels) has origin at bottom left.
        // Image formats on the other hand usually deal with top right.
        for y in (0..self.height as usize).rev() {
            data.extend(
                reader
                    .iter()
                    .skip(y * self.width as usize)
                    .take(self.width as usize)
                    .flatten(),
            );
        }
        Ok(data)
    }
```
    
Is there anything above that we're not releasing??

**Comment by nobbele** [comment]
Can confirms it leaks on latest devel

**Comment by fairbairn** [comment]
Thank you for confirming, @nobbele 

If/when we get it resolved, would it be possible to patch version 5 as well?

We have our code base running against both versions, but version 6.0.rc1 acts a good bit differently when it comes to that particular function call.  It must lock something up in version 6 vs version 5 because our fps(ctx) value drops in half, from 60 fps to 30 fps.

Not sure why, but it impacts our application.

**Comment by nobbele** [comment]
That sounds like a weird problem, open another issue for that?

**Comment by fairbairn** [comment]
Yes, we'll open up another issue for that.

**Comment by fairbairn** [comment]
Hi @nobbele 

We pulled the latest master (rc2) today and went through the process again.

Some interesting developments that I hope will shed some light on the issue and maybe a resolution.

First of all, in context, when running v5 of ggez, we could run at 60fps without an issue, except for the memory leak.

Unfortunately, when upgrading to 6.rc2, we've encountered two situations that vary slightly...

From the code below...

```
    pub fn capture_stage(&mut self, ctx: &mut Context, canvas: &graphics::Canvas) {
        let data = canvas.raw_image().to_rgba8(ctx).unwrap();
        // let data = canvas.to_rgba8(ctx).unwrap();
    }
```

We had originally attempted to use the suggested `canvas.to_rgba8()` function, however, because it internally flushes and performs the flip, it buries our FPS rate almost immediately, dropping down to as low as 17fps.  It basically can't keep up.

Keep in mind, the prior function we used in v5..

```
        let screen = canvas.image().to_rgba8(ctx).unwrap();    // <<<<<<<<<<<<<<<<<<<<<<<<< THIS CAUSES A LEAK
```

caused a leak, but did perform.

So in the new code above, we could not use that, and had to go direct to the exposed (and new) raw_image instead and perform the same operation... which does work, but interestingly, although the FPS will start at 60, it will drop steadily downward until around 24 fps or so and then go between 24 and 30, but never getting up into the 50s or 60 again.

Our particular application performs a screen capture of the canvas we're pre-drawing on (as an off-screen canvas) at 30fps, so every 33ms, we perform the capture...

So our draw function looks like this...

```
    fn draw(&mut self, ctx: &mut Context) -> GameResult {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last_time);

        if elapsed.as_millis() >= 32 {
            self.clear(ctx);
            self.draw(ctx);
            self.capture(ctx);
            self.last_time = now;
            self.frame += 1;
        } else {
            timer::yield_now();
        }

        if self.frame % 30 == 0 {
            println!(
                "FPS: {}, Elapsed: {}",
                ggez::timer::fps(ctx),
                elapsed.as_millis()
            );
        }

        self.render(ctx);
        graphics::present(ctx).unwrap();
        Ok(())
    }
```

why track the time ourselves?  well, we also tested when the machine goes to sleep and then wakes up, and we have to adjust our timing (because we're doing audio and video sync things), and if that happens, ggez (in version 5) would never come back from its timer check_update_time function... it's as if it lost track and hung, so we couldn't use that...

This works above, in v5 and v6, but the canvas capture (and our canvas is 1920 x 1080 in size) is bogging out the system now.

We can perform the vertical flip on our end, which is why we can get away with the raw_image() path, but obviously it would be ideal if this was all done performant with a proper call which insured that we're getting a clean capture of the canvas.

Memory leak is still present in latest rc2/master submission.

We ideally should be able to capture the canvas at 60fps... and maybe even opt out of the flip to buy back some speed, but something else is going on, and I think it relates to the leak as well - it's as if some kind of gpu back-end level lock of some kind is getting into the mix, and it doesn't get released (memory leak?) and it's slowing down captures over time.

If there's some other way to get to the low-level data behind the canvas for our high-speed capture purposes, I'll all ears.

Thanks.

**Comment by fairbairn** [comment]
Hi @nobbele 

Update since upgrading to rc2/master push yesterday.

So we're still encountering the memory leak, but we're now encountering a gradual slowdown in overall FPS (reported by fps(ctx) as run time extends...

First of all, we had updated our capture to the following...


```
    pub fn capture_stage(&mut self, ctx: &mut Context, canvas: &graphics::Canvas) {
        let data = canvas.raw_image().to_rgba8(ctx).unwrap();
        // let data = canvas.to_rgba8(ctx).unwrap();
  .....
    }

```

We had hoped to utilize the new native canvas.to_rgba8() function, but it causes an immediate FPS slowdown, like below 30fps.  

We believe this is caused by the flushing and flip code.  We elected to perform the flip code on our end and instead use the raw_image().to_rgba8() function instead (bypassing the flush), which does help (initially), but over time, the FPS continues to drop...

It can drop as low as 17fps and then gradually move up to 24fps and drop back down.

Nothing else has changed, and the system ran at 59-60fps previously in v5.

I believe they are connected....the memory leak is likely something failing to get released at the lower levels of the backend, and this possibly holds onto some resource or even a temporal lock of some kind which gradually impacts the performance of the entire system.

We would like to stick with rc2 as we've upgraded all of the code from v5, but we're uncertain how to move forward with the gradual degradation.

Unfortunately for our application, we need to be able to capture the 1080p canvas at 30fps or better while keeping the overall processing rate around 55-60fps.

Here is our draw function...

```
    fn draw(&mut self, ctx: &mut Context) -> GameResult {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last_time);

        if elapsed.as_millis() >= 32 {
            self.clear(ctx);
            self.draw(ctx);
            self.capture(ctx);
            self.last_time = now;
            self.frame += 1;
        } else {
            // timer::yield_now();
        }

        if self.frame % 30 == 0 {
            println!(
                "FPS: {}, Elapsed: {}",
                ggez::timer::fps(ctx),
                elapsed.as_millis()
            );
        }

        self.render(ctx);
        graphics::present(ctx).unwrap();
        Ok(())
    }

```

We're using our own timing for the draw and capture at 30fps intervals because when our system goes to sleep and is then restored, the timer check_update_time hangs... we're tracking accurate time instead of that seems to work after the wake up.

Let me know if we can be of further assistance in resolving this issue.

**Comment by fairbairn** [comment]
Another update, we backported our 6.rc2 code BACK to v5, had to change the capture to this line instead..

```
    pub fn capture_stage(&mut self, ctx: &mut Context, canvas: &graphics::Canvas) {
        let data = canvas.image().to_rgba8(ctx).unwrap();
    }
```

If that is in there, it degrades quickly from 60fps down to 24fps and even lower...in like 30 seconds of run-time.
If we remove that one line, it holds at 60fps.
And this is v5, so the problem persists across both v5 and master (v6.rc2).

Is it possible to capture a 1080p canvas at 60fps?  (we could live with 30fps, which is how we have it coded, but we want the update() and draw() routines to still run at 60fps).

Anyway, at least we know what we're looking for now... it's clearly something within the Image::to_rgba8() routine.

Thanks again.

**Comment by fairbairn** [comment]
Another update... I do not believe the frame rate degradation is due to that to_rgba8() call.  We removed our drawing code entirely, leaving capture, and we can sustain 60fps.

We'll continue to monitor the leak issue on future pulls.

**Comment by nobbele** [comment]
It's very possible this is an issue with gfx pre-ll which isn't worked on anymore. There are plans to rewrite graphics to something else though, but this issue might not be fixed for quite some time.

**Comment by fairbairn** [comment]
Very possible.

I guess if it continues, we'll have to rewrite our core engine around something else (possibly wgpu).  Might be something we can live with for a window of time - if we're losing 3MB a minute, then basically 24 hours is going to cost us 4.3GB of RAM in leakage.  Thinking after 24 hours we could restart our process.  Not ideal, but isn't the end of the world.

I have a question while I have you...(just requesting some insight)...

We're drawing about as fast as we can get right now with ggez, but we need rounded corners on our images, and the images often scale down into the canvas.

If we add a shader in GLSL to perform the rounded corners (opacity mask) but also provide a non-1.0 scale on DrawParams, will it run the shader first and then scale, or scale first and then run the shader?

Our assumption is that it would run the shader and then scale, which would not work for us, because it would then change the border size overall, but wanted to ask.

If it does execute the shader before the scale, then we'd have to set scale to 1.0, 1.0 and instead perform a resize within our shader along with the rounded border logic.

So if that's the scenario, would a single shader have to do it all, or could we chain two shaders together to execute at the time of the draw?

How would you suggest we tackle the problem?

**Comment by PSteinhaus** [comment]
> We believe this is caused by the flushing and flip code. We elected to perform the flip code on our end and instead use the raw_image().to_rgba8() function instead (bypassing the flush), which does help (initially), but over time, the FPS continues to drop...

`to_rgba8` always creates a GPU buffer for the image and a tiny command buffer, tells it to copy texture data to the new buffer, flushes it, and then reads from that new buffer, waiting for all the data from the GPU to arrive. So you can't really get around doing at least one flush. If you go this way though, you will always only get the content the canvas had in the previous frame, since without the flush in `Canvas::to_rgba8` the new content will not be rendered yet.

> Is it possible to capture a 1080p canvas at 60fps? (we could live with 30fps, which is how we have it coded, but we want the update() and draw() routines to still run at 60fps).

Technically it should be possible, but it's not what ggez is built for, so you might need to work with gfx directly.
[`ImageGeneric` exposes its texture](https://docs.rs/ggez/0.6.0-rc1/ggez/graphics/struct.ImageGeneric.html#impl), so it should be possible to copy it to a GPU buffer (similarly to what `to_rgba8` does internally). To implement that you'd probably need to get the current gfx encoder via [`graphics::gfx_objects`](https://docs.rs/ggez/0.6.0-rc1/ggez/graphics/fn.gfx_objects.html).

May I ask why you need to make snapshots of a canvas all the time? Do you actually need them in main memory for something, or would it be possible to keep them on the GPU inside some kind of buffer? If that was possible you should probably be able to fix your performance in a somewhat straight way. If not then I'd need more information about what you want to do before making any suggestions.

A simple thing that may or may not be applicable to your situation is going with `graphics::screenshot` instead to screenshot the screen directly, without using any kind of canvas. This would probably only make a difference though if it allowed you to completely stop using the off screen canvas, so it depends upon whether you actually need the canvas for something else as well.

**Comment by fairbairn** [comment]
Thank you @PSteinhaus 

I understand the flush.  If it performs, we'll use it.  So far, it has a significant impact for our use-case, likely because we're so time sensitive.

We need to capture the frame at 30fps because we're creating a scene in real-time and capturing the frames which go into a video encoder, hence the performance target.

We tried `screenshot` but it suffered from the same hit (same flush), so we did not benefit.

**Comment by fairbairn** [comment]
@nobbele we got rounded corners to work, found the new mesh function for rounded rectangle and by changing up colors and blend modes, we achieved our goal without resorting to GLSL.

**Comment by PSteinhaus** [comment]
> we got rounded corners to work, found the new mesh function for rounded rectangle and by changing up colors and blend modes, we achieved our goal without resorting to GLSL.

I bet he'll be happy to hear that; him being to the one who added them, if I remember correctly.

> We need to capture the frame at 30fps because we're creating a scene in real-time and capturing the frames which go into a video encoder, hence the performance target.

Then all I can suggest is using `graphics::gfx_objects` to create something like `graphics::screenshot`, but adapted to your usecase:

- to avoid a second flush you'd need to put the `copy_texture_to_buffer_raw` at the end of your draw function, just before calling `present`
- instead of recreating the download buffer every frame create it just once and read from it after `present` has been called (and only recreate it if the window size changes (but I guess that'd screw up video encoding anyway, so maybe don't))

Looking at this naively I guess that it's the download buffer that's leaking. It would make sense, since the original gfx example using `create_download_buffer` doesn't call it every (second) frame, recreating it all the time, so it being leaky might have been missed.

But then again, that might be wishful thinking, since it would allow you (and also us) to work around it easily by just keeping a handle to a screenshot download buffer somewhere in the graphics context (or in your case somewhere in the main state struct of your application).

**Comment by fairbairn** [comment]
Great, thanks for the hints.

We're actually drawing and capturing from a separate canvas, in preparation for it to run headless.

Before the present, we draw our canvas to the None version and draw to our visible window for feedback during the development process, but in the final deployment, there will be no window.

Your hints may help us tune it tighter in the interim, so THANK YOU!

I think after we get through v1, we'll re-assess ggez vs wgpu (although greatly prefer ggez) and figure out a path.

I'd pay someone to build me a slightly higher level (ggez-like) drawing api on top of wgpu, so if someone comes to mind, send them my way.

**Comment by nobbele** [comment]
I've been working a bit on a higher level wgpu renderer, but I don't if I will work on it to a usable state.

**Comment by fairbairn** [comment]
Well, maybe I could fund that effort ;)

**Comment by PSteinhaus** [comment]
> We're actually drawing and capturing from a separate canvas, in preparation for it to run headless.
> 
> Before the present, we draw our canvas to the None version and draw to our visible window for feedback during the development process, but in the final deployment, there will be no window.
> 
> Your hints may help us tune it tighter in the interim, ...

It should also help you tune it tighter in the final deployment, since you can use this method with an off-screen canvas as well as with the visible screen.

**Comment by PSteinhaus** [comment]
> Looking at this naively I guess that it's the download buffer that's leaking.

> But then again, that might be wishful thinking, since it would allow you (and also us) to work around it easily by just keeping a handle to a screenshot download buffer somewhere in the graphics context... 

Ok, I just tried that out and sadly it doesn't seem to be the download buffer that's leaking, at least not it as a whole.
Creating the buffer just once (recreating it whenever the demanded size changes) doesn't fix the memory leak.

**Comment by PSteinhaus** [comment]
Haha! I found it!

It's the command buffer that's leaking.
Probably it's not a good idea to create multiple of them.

changing this:
```rust
let factory = &mut *gfx.factory;
let mut local_encoder = GlBackendSpec::encoder(factory);
```
to this:
```rust
//let factory = &mut *gfx.factory;
//let mut local_encoder = GlBackendSpec::encoder(factory);
let local_encoder = &mut gfx.encoder;
```
fixes the leak

Now the question is whether it might have any drawbacks to use the global command buffer instead of creating new ones.
The only thing that comes to my mind is that this could cause flushes of the global command buffer before the user actually wanted to flush it.

But would that actually ever be a problem?
`to_rgba8` needs a flush, so there'll always be 2 flushes, however you do it.
So the only remaining question is: Could something actually break if the command buffer got flushed more often? I don't think so, so I guess doing this is pretty safe and a perfect fix for the leak.

**Comment by PSteinhaus** [comment]
@fairbairn For you specific problem it might still be best to use `graphics::gfx_objects` to recreate `to_rgba8` for your use case, as described, since it will still save you a flush.

But apart from that, once the PR is merged, your initial approach should work just fine on devel.
Since I also added a persistent download buffer to the gfx context performance may or may not have improved for you. I'd be very interested to hear some feedback on that.

**Comment by fairbairn** [comment]
We'll test it today.  You rock.
